src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.Escape.cs

```diff
@@ -0,0 +1,185 @@
+        {
+            for (int i = 0; i < input.Length; i++)
+            {
+                if (RegexCharCategory.IsMetachar(input[i]))
+                {
+                    return EscapeImpl(targetSpan, input, i, destination, out charsWritten, out spanSuccess);
+                }
+            }
+
+            // If nothing to escape, return the input.
+            return input.CopyInput(targetSpan, destination, out charsWritten, out spanSuccess);
+        }
+
+        private static string EscapeImpl(bool targetSpan, ReadOnlySpan<char> input, int i, Span<char> destination, out int charsWritten, out bool spanSuccess)
+        {
+            // For small inputs we allocate on the stack. In most cases a buffer three 
+            // times larger the original string should be sufficient as usually not all 
+            // characters need to be encoded.
+            // For larger string we rent the input string's length plus a fixed 
+            // conservative amount of chars from the ArrayPool.
+            Span<char> buffer = input.Length <= (EscapeMaxBufferSize / 3) ? stackalloc char[EscapeMaxBufferSize] : default;
+            ValueStringBuilder vsb = !buffer.IsEmpty ?
+                new ValueStringBuilder(buffer) :
+                new ValueStringBuilder(input.Length + 200);
+
+            char ch = input[i];
+            vsb.Append(input.Slice(0, i));
+
+            do
+            {
+                vsb.Append('\\');
+                switch (ch)
+                {
+                    case '\n':
+                        ch = 'n';
+                        break;
+                    case '\r':
+                        ch = 'r';
+                        break;
+                    case '\t':
+                        ch = 't';
+                        break;
+                    case '\f':
+                        ch = 'f';
+                        break;
+                }
+                vsb.Append(ch);
+                i++;
+                int lastpos = i;
+
+                while (i < input.Length)
+                {
+                    ch = input[i];
+                    if (RegexCharCategory.IsMetachar(ch))
+                        break;
+
+                    i++;
+                }
+
+                vsb.Append(input.Slice(lastpos, i - lastpos));
+            } while (i < input.Length);
+
+            return vsb.CopyOutput(targetSpan, destination, out charsWritten, out spanSuccess);
+        }
+
+        /// <summary>
+        /// Unescapes all metacharacters (including (,),[,],{,},|,^,$,*,+,?,\, spaces and #)
+        /// </summary>
+        private static string Unescape(bool targetSpan, ReadOnlySpan<char> input, Span<char> destination, out int charsWritten, out bool spanSuccess)
+        {
+            for (int i = 0; i < input.Length; i++)
+            {
+                if (input[i] == '\\')
```

Comment:
Is it faster if you use the IndexOf that's in MemoryExtensions.cs which goes into the vectorized implementation in SpanHelpers?

If the string is small (and there's often a hit early) it may be slower, but we figured it was good enough for String.IndexOf.